---
title: "Class 14: Pathway Analysis from RNA-Seq Results"
author: "Clarissa Savko (PID:  A69028482)"
format: pdf
toc: true
---
```{r}
#| message : false
#| warning : false
library(DESeq2)
```
# Data Import
```{r}
metaFile <- "GSE37704_metadata.csv"
countFile <- "GSE37704_featurecounts.csv"

# Import metadata and take a peak
colData = read.csv(metaFile, row.names=1)
head(colData)
```
# Data Tidying
```{r}
countData = read.csv(countFile, row.names=1)
head(countData)
```
Q. Complete the code below to remove the troublesome first column from countData

```{r}
countData <- as.matrix(countData[,-1])
head(countData)
```

Q. Complete the code below to filter countData to exclude genes (i.e. rows) where we have 0 read count across all samples (i.e. columns).

How many genes do we have to start with?

```{r}
nrow(countData)
```
1) Find the rowSums() this will be zero for any genes with no count data
2) Find the zero sum genes
3) Remove them

```{r}
to.rm.inds <- rowSums(countData) == 0 
countData <- countData[!to.rm.inds,]
nrow(countData)
```


# DESeq setup and analysis

```{r}
dds <- DESeqDataSetFromMatrix(countData = countData, 
                              colData = colData,
                              design= ~condition)
```

```{r}
dds = DESeq(dds)
```

# Save Results

Q. Call the summary() function on your results to get a sense of how many genes are up or down-regulated at the default 0.1 p-value cutoff.

```{r}
res = results(dds)
summary(res)
```
# Visualize
```{r}
plot( res$log2FoldChange, -log(res$padj) )
```

```{r}
pc <- prcomp(t(countData), scale= T)
summary(pc)
```
```{r}
plot(pc$x[,1], pc$x[,2], col=as.factor(colData$condition), pch=15)
```
Q. Improve this plot by completing the below code, which adds color and axis labels



```{r}
# Make a color vector for all genes
mycols <- rep("gray", nrow(res) )

# Color red the genes with absolute fold change above 2
mycols[ abs(res$log2FoldChange) > 2 ] <- "red"

# Color blue those with adjusted p-value less than 0.01
#  and absolute fold change more than 2
inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

plot( res$log2FoldChange, -log(res$padj), col= mycols, xlab="Log2(FoldChange)", ylab="-Log(P-value)" )
```
# Annotation Data

Q. Use the mapIDs() function multiple times to add SYMBOL, ENTREZID and GENENAME annotation to our results by completing the code below.

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")

columns(org.Hs.eg.db)

res$symbol = mapIds(org.Hs.eg.db,
                    keys= row.names(res), 
                    keytype="ENSEMBL",
                    column= "SYMBOL",
                    multiVals="first")

res$entrez = mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")

res$name =   mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype= "ENSEMBL",
                    column= "GENENAME",
                    multiVals="first")

head(res, 10)
```
Q. Finally for this section let's reorder these results by adjusted p-value and save them to a CSV file in your current project directory.

```{r}
res = res[order(res$pvalue),]
write.csv(res, file = "deseq_results.csv")
```

# Geneset enrichment/pathway analysis

```{r}
#| message: false
library(gage)
library(gageData)
library(pathview)
```

The `gage()` function wants a "vector of importance" in our case here it will be fold-change values with associated entrez gene names.

```{r}
foldchange <- res$log2FoldChange
names(foldchange) = res$entrez
```

```{r}
data(kegg.sets.hs)
keggres = gage(foldchange, gsets= kegg.sets.hs)
head(keggres$less)
```

hsa04110 cell cycle

```{r}
pathview(gene.data= foldchange, pathway.id = "hsa04110")
```
Have a look at my figure (@fig-cellcycle)

![Cell cycle hsa04110](hsa04110.pathview.png){#fig-cellcycle}
Q. Can you do the same procedure as above to plot the pathview figures for the top 5 down-reguled pathways?

```{r}
pathview(gene.data= foldchange, pathway.id = "hsa03030")
pathview(gene.data= foldchange, pathway.id = "hsa05130")
pathview(gene.data= foldchange, pathway.id = "hsa03013")
pathview(gene.data= foldchange, pathway.id = "hsa03440")

```

![DNA replication hsa03030](hsa03030.pathview.png)
![Pathogenic Escherichia coli infection hsa05130](hsa05130.pathview.png)
![RNA transport hsa03013](hsa03013.pathview.png)
![Homologous recombination hsa03440](hsa03440.pathview.png)


## Gene Ontology

```{r}
data(go.sets.hs)
data(go.subs.hs)
gobpsets = go.sets.hs[go.subs.hs$BP]
gobpres = gage(foldchange, gsets=gobpsets, same.dir=TRUE)

lapply(gobpres, head)
```

## Reactome

We will use the online version of Reactome. It wants a list of your genes. We will write this out from R here:

```{r}
sig_genes <- res[res$padj <= 0.05 & !is.na(res$padj), "symbol"]
print(paste("Total number of significant genes:", length(sig_genes)))

```

```{r}
write.table(sig_genes, file="significant_genes.txt", row.names=FALSE, col.names=FALSE, quote=FALSE)
```


Have a look at my figure (@fig-SMAD)

![SMAD signaling](R-HSA-2173793.png){#fig-SMAD}
Have a look at my figure (@fig-cellcycle)

![Cell cycle](R-HSA-1640170.png){#fig-cellcycle}
Q: What pathway has the most significant “Entities p-value”? Do the most significant pathways listed match your previous KEGG results? What factors could cause differences between the two methods?

Mitotic cell cycle. This matches the most downregulated pathway in the KEGG results. They are using different databases as references so they will show different results.

